### ISSUE TITLE
Data Link "__all_variables" not working properly

### ISSUE THREAD

**Issue by Ultraseamus** [grafana/grafana#37282]
**What happened**:

Using "__all_variables" in a data link on a repeating panel injects the current panel's value for the repeating variable.

**What you expected to happen**:

"__all_variables" is defined as your current dashbaord's variable selection. It should be the variable values you currently have selected.

**How to reproduce it (as minimally and precisely as possible)**:

Create a custom template variable with valid fields for a terms aggregation and an "All" value.

Select the "All" value.

Create a repeating panel with that variable as your field value for a Terms aggregation.

Create a Data Link that links to your current dashboard, and uses "__all_variables".

Click that data link in one of your repeated panels, notice that your custom variable will be set to the repeated value of that panel.

**Anything else we need to know?**:

Essentially, Data Links are broken when used in repeated panels, the panel you click the link on will become the only panel in the series. And there is no way to avoid it. That I know.

I believe this happens because "scopedVars" is being used for two different purposes. It is used to track the state of dashboard variables, and it is also use to track the local value of a repeated panel's variable. While the data link should be passed in a "scopedVars" that represents the dashboard's state, it instead gets the repeated panel's version.

**Environment**:
- Grafana version: 8.0.6
- Data source type & version: Elasticsearch

**Comment by Ultraseamus** [comment]
I've tested a fix in my dev environment that seems to work. I'd be happy to create a PR for it, but I'm curious to know what people think of my approach (it feels a little hacky).

Basically, in "replaceVariables" I check to see if an "__all_variables" value was passed in (which, at that point, already has all of the variables formatted for the URL, since it was passed in by getLinkSupplier). If it was, then I use that rather than trying to recreate it locally. The logic to create the URL formatted "all variables" is run twice in a row. Once in getLinkSupplier (done correctly) and once in replaceVariables (incorrect because it considers the panel's scopeVars).

`const variablesQuery = vars?.__all_variables
      ? vars?.__all_variables.value
      : urlUtil.toUrlParams(allVariablesParams);
    const timeRangeUrl = urlUtil.toUrlParams(getTimeSrv().timeRangeForUrl());`

**Comment by axelavargas** [comment]
Hello @Ultraseamus, 👋🏾  

Thank you for reporting this issue, would you mind sharing your panel.json? It would help us reproduce this scenario and triage the issue properly. 

Regarding opening a PR, yes! it's awesome you want to contribute 🎉, I would say we can first confirm the issue (with your panel.json) and afterward you can start working on the fix.

**Comment by Ultraseamus** [comment]
@axelavargas Sure.

The issue can be reproduced on https://play.grafana.org/, I've reproduced it there and add the panel JSON below.

In this dashboard, "$customvar" is  Custom variable with two options: "@source" and "@hostname". When I extracted the panel JSON, "$customvar" was set to all. And you can see that, since it is a repeating panel, it is storing its repeated value for "$customvar". This is good for repeating support, but bad for the "__all_variables" option.

The data Link in this panel should do nothing. It linked to the current dashboard I was using, and the only variable it defined was "__all_variables" which should have been the exact values already set. However, when you click on this datalink "$customvar" is set to "@source".


`{
  "id": 83,
  "gridPos": {
    "x": 0,
    "y": 1,
    "w": 12,
    "h": 8
  },
  "type": "timeseries",
  "title": "Panel Title",
  "scopedVars": {
    "customvar": {
      "text": "@source",
      "value": "@source",
      "selected": false
    }
  },
  "repeat": "customvar",
  "repeatDirection": "h",
  "datasource": "Elastic Metrics",
  "targets": [
    {
      "refId": "A",
      "query": "",
      "alias": "",
      "metrics": [
        {
          "type": "count",
          "id": "1"
        }
      ],
      "bucketAggs": [
        {
          "id": "3",
          "type": "terms",
          "settings": {
            "min_doc_count": "0",
            "size": "10",
            "order": "desc",
            "orderBy": "_term"
          },
          "field": "$customvar"
        },
        {
          "type": "date_histogram",
          "id": "2",
          "settings": {
            "interval": "auto"
          },
          "field": "@timestamp"
        }
      ],
      "timeField": "@timestamp"
    }
  ],
  "options": {
    "tooltip": {
      "mode": "single"
    },
    "legend": {
      "displayMode": "list",
      "placement": "bottom",
      "calcs": []
    }
  },
  "fieldConfig": {
    "defaults": {
      "custom": {
        "drawStyle": "line",
        "lineInterpolation": "linear",
        "barAlignment": 0,
        "lineWidth": 1,
        "fillOpacity": 0,
        "gradientMode": "none",
        "spanNulls": false,
        "showPoints": "auto",
        "pointSize": 5,
        "stacking": {
          "mode": "none",
          "group": "A"
        },
        "axisPlacement": "auto",
        "axisLabel": "",
        "scaleDistribution": {
          "type": "linear"
        },
        "hideFrom": {
          "tooltip": false,
          "viz": false,
          "legend": false
        },
        "thresholdsStyle": {
          "mode": "off"
        }
      },
      "color": {
        "mode": "palette-classic"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "value": null,
            "color": "green"
          },
          {
            "value": 80,
            "color": "red"
          }
        ]
      },
      "mappings": [],
      "links": [
        {
          "title": "Link which should do nothing",
          "url": "d/YI95GyqMz?${__all_variables}"
        }
      ]
    },
    "overrides": []
  }
}`

**Comment by tskarhed** [comment]
Reproduced [here](https://play.grafana.org/d/zbQUynM7k/data-link-__all_variables-not-working-properly-37282?orgId=1&var-customvar=All)

**Comment by axelavargas** [comment]
Thank you @tskarhed 👍🏾,  @Ultraseamus If you are still up for it, you can open the PR, we will be happy to review it :)

**Comment by grafanabot** [comment]
We've closed this issue since it needs more information and hasn't had any activity recently. We can re-open it after you you add more information. To avoid having your issue closed in the future, please read our [CONTRIBUTING](https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md) guidelines.

Happy graphing!

**Comment by Ultraseamus** [comment]
@hugohaggmark: The bug is specifically around "${__all_variables}" not working properly.

The way "${__all_variables}" should work is for it to take your current dashboard's variable values and add them into your url. Which is particularly handy when you are data linking to the dashboard you are already on.  A user would expect to be able to click on a single dimension in a chart and see that one dimension be updated without any other selections they've made being erased.


To make a better example from what I explained above. Let's say you have two template variables:

Region: (All, West, East, North, South)
SKU: (All, SKU_1, SKU_2, SKU_3)

You have both Region and SKU set to "All".

You make a repeating panel based on Region which splits out to 4 panels. For that repeating series you want to list the different server SKUs in each region.

Then we create our data link. Our intention is to allow users to click on a SKU row in any of the repeated panels and have the dashboard filter to that SKU. This should be a good use for "${__all_variables}" because we want any variable selections the user has made up to that point to stay selected. We don't want them to be passed in empty and revert to defaults.

But, if we use "${__all_variables}" it will not actually fill out the URL will all currently selected variables. Instead it will change "Region" from the selected value "All" to whichever repeated instance of Region that belong to the panel you clicked. Which means the variable you are repeating on will move off of All and onto a single value, which will force your repeated series to reduce.

There is no way to use "${__all_variables}" within a repeated panel without this happening. "${__all_variables}" no longer means all currently selected variables because it applies that panel's "ScopedVars" to it.


@tskarhed: reproduced this behavior in that link, but I guess all elasticsearch datasources are broken on the site at the moment?

But it is easy to reproduce the issue with the steps I originally described. And it is definitely broken in its current state. Or, "${__all_variables}" needs to be redefined in the documentation (it would still be not very useful for repeating panels, but then by design, I guess :)).


I got side-tracked over the past couple months, but if it is agreed that this is an issue, I can still look at making the PR I described above. The fix is pretty simple.
